packages/runtime/container-runtime/src/compatUtils.ts

```diff
@@ -0,0 +1,200 @@
+	 */
+	modernConfig: IContainerRuntimeOptionsVersionDependent[K];
+}
+
+/**
+ * Mapping of version-dependent options to their compatibility related configs.
+ */
+const versionDependentOptionConfigMap: {
+	[K in keyof IContainerRuntimeOptionsVersionDependent]: IVersionDependentOptionConfig<K>;
+} = {
+	enableGroupedBatching: {
+		minVersionForModernConfig: "2.0.0",
+		legacyConfig: false,
+		modernConfig: true,
+	},
+	compressionOptions: {
+		minVersionForModernConfig: "2.0.0",
+		legacyConfig: disabledCompressionConfig,
+		modernConfig: enabledCompressionConfig,
+	},
+
+	enableRuntimeIdCompressor: {
+		// We do not yet want to enable idCompressor by default since it will increase bundle sizes,
+		// and not all customers will benefit from it. Therefore, we will require customers to explicitly
+		// enable it. We are keeping it as a version-dependent option as this may change in the future.
+		minVersionForModernConfig: undefined,
+		// Additionally, for IdCompressorMode `undefined` represents a logical state (off). However, to satisfy the Required<>
+		// constraint we need to have it defined, so we trick the type checker here.
+		legacyConfig: undefined as unknown as "on" | "delayed",
+		modernConfig: "on",
+	},
+	explicitSchemaControl: {
+		// This option is unique since it was actually introduced before 2.0.0, but its purpose is to prevent 1.x clients from
+		// joining a session. Therefore, we will have it be `true` when the compatibility mode is set to >=2.0.0 and we do not
+		// want any 1.x clients to join.
+		minVersionForModernConfig: "2.0.0",
+		legacyConfig: false,
+		modernConfig: true,
+	},
+	flushMode: {
+		// Note: 1.x clients are compatible with TurnBased flushing, but here we elect to remain on Immediate flush mode
+		// as a work-around for inability to send batches larger than 1Mb. Immediate flushing keeps batches smaller as
+		// fewer messages will be included per flush.
+		minVersionForModernConfig: "2.0.0",
+		legacyConfig: FlushMode.Immediate,
+		modernConfig: FlushMode.TurnBased,
+	},
+	gcOptions: {
+		// Explicitly disable running Sweep in compat mode "2". Although sweep is supported in 2.x, it is disabled by default.
+		// This setting explicitly disables it to be extra safe.
+		minVersionForModernConfig: "3.0.0",
+		legacyConfig: {},
+		modernConfig: { enableGCSweep: true },
+	},
+};
+
+/**
+ * Returns the default version-dependent configs for a given compatibility mode.
+ */
+export function getConfigsForCompatMode(
+	compatibilityMode: Required<IContainerRuntimeOptionsInternal>["compatibilityMode"],
+): IContainerRuntimeOptionsVersionDependent {
+	const defaultConfigs = {
+		gcOptions: {},
+		flushMode: FlushMode.TurnBased,
+		compressionOptions: enabledCompressionConfig,
+		enableRuntimeIdCompressor: undefined as unknown as "on" | "delayed",
+		enableGroupedBatching: true,
+		explicitSchemaControl: false,
+	};
```

Comment:
Can you declare this outside the function and provide justification for why values are the way they are? It looks like you cannot currently compute this easily from the data defined above because we have up-to-now been inconsistent about how we introduce new options, but hopefully in the future that could be changed.